libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,176 @@
+  getScriptToRun
+} from './ShellOperationRunnerPlugin';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStrings = {
+  SHARD_INDEX: '{shardIndex}',
+  SHARD_COUNT: '{shardCount}'
+} as const;
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStringRegex = {
+  SHARD_INDEX: new RegExp(TemplateStrings.SHARD_INDEX, 'g'),
+  SHARD_COUNT: new RegExp(TemplateStrings.SHARD_COUNT, 'g')
+} as const;
+
+/**
+ * Phased command that shards a phase into multiple operations.
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { rushConfiguration } = context;
+
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project, settings: operationSettings } = operation;
+    if (phase && project && operationSettings?.sharding && !operation.runner) {
+      const { count: shards, shardScriptConfiguration } = operationSettings.sharding;
+
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project,
+        settings: operationSettings
+      });
+      existingOperations.add(collatorNode);
+
+      const parentFolderFormat: string =
+        operationSettings.sharding.outputFolderArgument?.parentFolderName ??
+        `.rush/operations/${phase.logFilenameIdentifier}/shards`;
+
+      const parentFolder: string = parentFolderFormat;
+
+      const collatorDisplayName: string = `${getDisplayName(phase, project)} - collate`;
+      const collatorRawScript: string | undefined = getScriptToRun(project, phase.name, undefined);
+
+      if (collatorRawScript === undefined && phase.missingScriptBehavior === 'error') {
+        throw new Error(
+          `The project '${project.packageName}' does not define a '${phase.name}' command in the 'scripts' section of its package.json`
+        );
+      }
+
+      const customParameters: readonly string[] = getCustomParameterValuesForPhase(phase);
+      if (collatorRawScript) {
+        const collatorRunner: ShellOperationRunner = new ShellOperationRunner({
+          commandToRun: formatCommand(collatorRawScript, customParameters),
+          displayName: collatorDisplayName,
+          phase,
+          rushConfiguration,
+          rushProject: project,
+          environment: {
+            ...process.env,
+            [EnvironmentVariableNames.RUSH_SHARD_PARENT_FOLDER]: parentFolder,
+            [EnvironmentVariableNames.RUSH_SHARD_COUNT]: shards.toString()
```

Comment:
Seems slightly odd that these are passed via env vars, but other values are passed via CLI.